src/DataProtection/DataProtection/src/KeyManagement/KeyRingBasedDataProtector.cs

```diff
@@ -381,19 +359,55 @@ public byte[] GetAadForKey(Guid keyId, bool isProtecting)
             }
         }
 
-        private sealed class PurposeBinaryWriter : BinaryWriter
+        internal static byte[] BuildAadTemplateBytes(string[] purposes)
         {
-            public PurposeBinaryWriter(MemoryStream stream) : base(stream, EncodingUtil.SecureUtf8Encoding, leaveOpen: true) { }
+            // additionalAuthenticatedData := { magicHeader (32-bit) || keyId || purposeCount (32-bit) || (purpose)* }
+            // purpose := { utf8ByteCount (7-bit encoded) || utf8Text }
+
+            var keySize = sizeof(Guid);
+            int totalPurposeLen = 4 + keySize + 4;
 
-            // Writes a big-endian 32-bit integer to the underlying stream.
-            public void WriteBigEndian(uint value)
+            var purposeLengthsPool = ArrayPool<int>.Shared.Rent(purposes.Length);
+            for (int i = 0; i < purposes.Length; i++)
             {
-                var outStream = BaseStream; // property accessor also performs a flush
-                outStream.WriteByte((byte)(value >> 24));
-                outStream.WriteByte((byte)(value >> 16));
-                outStream.WriteByte((byte)(value >> 8));
-                outStream.WriteByte((byte)(value));
+                string purpose = purposes[i];
+
+                int purposeLength = EncodingUtil.SecureUtf8Encoding.GetByteCount(purpose);
+                purposeLengthsPool[i] = purposeLength;
+
+                var encoded7BitUIntLength = purposeLength.Measure7BitEncodedUIntLength();
+                totalPurposeLen += purposeLength /* length of actual string */ + encoded7BitUIntLength /* length of 'string length' 7-bit encoded int */;
             }
+
+            byte[] targetArr = new byte[totalPurposeLen];
+            var targetSpan = targetArr.AsSpan();
+
+            // index 0: magic header
+            BinaryPrimitives.WriteUInt32BigEndian(targetSpan.Slice(0), MAGIC_HEADER_V0);
+            // index 4: key (skipped for now, will be populated in `GetAadForKey()`)
+            // index 4 + keySize: purposeCount
+            BinaryPrimitives.WriteInt32BigEndian(targetSpan.Slice(4 + keySize), purposes.Length);
+
+            int index = 4 + keySize + 4; // starting from first purpose
+            for (int i = 0; i < purposes.Length; i++)
+            {
+                string purpose = purposes[i];
+
+                // writing `utf8ByteCount (7-bit encoded integer) || utf8Text`
+                // we have already calculated the lengths of the purpose strings, so just get it from the pool
+                index += targetSpan.Slice(index).Write7BitEncodedInt(purposeLengthsPool[i]);
+
+#if NET10_0_OR_GREATER
```

Comment:
nit; since we have the array path *anyway*, I wonder if we should just use the second version in all cases, to avoid an unnecessary `#if` - it won't have any relevant perf delta - anything gained in one direction will be offset by things like the extra `Slice`